pybamm/expression_tree/broadcasts.py

```diff
@@ -417,15 +462,17 @@ class FullBroadcast(Broadcast):
 
     def __init__(
         self,
-        child,
-        broadcast_domain=None,
-        auxiliary_domains=None,
-        broadcast_domains=None,
-        name=None,
+        child_input: NumberType | float | pybamm.Symbol,
```

Comment:
sort of overlaps with `ChildValue` / `ChildSymbol` defined in binary ops file (except no ndarray, should this be here as well?). Can we just define this in one place and re-use everywhere?

Also, surely float is a `NumberType`?

Reply:
Types from `numbers` aren't supported for static type checking (at least in mypy) - they recommend hinting using `complex/float/int` instead according to https://peps.python.org/pep-0484/#the-numeric-tower.

If we want to keep using `numbers.Numbers` in functions like this one, the suggestion is to use `typing.SupportsFloat` (which is what `NumberType` is defined as at the top of the file), which just checks if the class has a `_float_` function (the accepted answer here was really useful to explain the vagaries https://stackoverflow.com/questions/69334475/how-to-hint-at-number-types-i-e-subclasses-of-number-not-numbers-themselv) - although if we start supporting complex numbers this would break.

What I haven't quite worked out is why mypy complains when type-hinting with `numbers.Number` in `FullBroadcast` & `FullBroadcastToEdges`, but not for `PrimaryBroadcast`, `Scalar`, and a few other functions...